packages/@vue/cli-service/lib/commands/serve.js

```diff
@@ -102,6 +102,7 @@ module.exports = (api, options) => {
     const protocol = useHttps ? 'https' : 'http'
     const host = args.host || process.env.HOST || projectDevServerOptions.host || defaults.host
     portfinder.basePort = args.port || process.env.PORT || projectDevServerOptions.port || defaults.port
```

Comment:
The basePort is passed as string type when using environment variable and dotenv.
Then the result is as follows.

```js
// PORT=3000 yarn serve
portfinder.highestPort  = portfinder.basePort + 1 // 30001
```

So you might explicitly cast it to number type.

```suggestion
     portfinder.basePort = Number(args.port || process.env.PORT || projectDevServerOptions.port || defaults.port)
```